Add onchange configuration for script engine

lkubb opened the pull request:
### What does this PR do?
Adds configuration for the script engine to only fire an event if the event tag-specific data changes compared to the last run (see the linked issue).

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62867

### Previous Behavior
Fires an event every time the script is run.

### New Behavior
Can be configured to watch for changes between runs.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

lkubb committed:
Add onchange configuration for script engine

lkubb committed:
Add more tests for script engine

Ch3LL reviewed (commented):
I would prefer to not use this function if we don't need to. I understand why you are doing that and ping @waynew for any ideas here.

waynew commented on salt/engines/script.py:
```diff
@@ -53,7 +63,7 @@ def _get_serializer(output):
 
 
-def start(cmd, output="json", interval=1):
+def start(cmd, output="json", interval=1, onchange=False):
```
Not positive @Ch3LL - shouldn't we default to True here to maintain backwards compatibility? If we want to change behavior in the future we can set this to None, and do `if onchanges is None: log.warning('onchanges was not set, defaulting to False but this will change after YYYY-MM-DD. To supress this warning set onchanges to True or False')`

lkubb replied:
Not sure I'm following, but `onchange=False` exihibits the same behavior as without this PR (firing events every time the script is run). The new feature is opt-in by setting `onchange: true`.

Ch3LL replied:
I agree that this does not seem to be a breaking change to me.

waynew commented on salt/engines/script.py:
```diff
@@ -132,3 +150,8 @@ def start(cmd, output="json", interval=1):
+
+
+# helper to test the start function
+def _running():
```
I have some ideas but I'm AFK at the moment. I'm pretty sure I can put together some super reasonable tests without needing this, though

lkubb replied:
In the meantime I refactored the script engine to make use of the solution I linked above, which feels a little less weird. Will push the commit now. I'm happy to learn though, so please feel free to point me in the right direction or take over the testing. :)

lkubb committed:
Refactor script engine for testability

Ch3LL commented:
Looks good to me. @waynew can you also review please

waynew reviewed (commented):
Well, I *thought* that was just a comment but apparently GitHub decided posting on the comment page should open up a review :facepalm:

waynew commented:
@lkubb so on today's Test Clinic I went ahead and added a fixture and a couple of tests. https://github.com/saltstack/salt/pull/62910 is my PR that adds the fixture, and you can find the tests I started on over here: https://github.com/saltstack/salt/compare/master...waynew:salt:engine-scripts-tests?expand=1

I think the timer approach is probably a reasonable way to go for this.

garethgreenaway commented:
@lkubb Making sure you saw the comments from @waynew above.

lkubb committed:
Revert "Refactor script engine for testability"

This reverts commit f75eb062fe4086ddd785efd55f46a48ec2d2c8d9.

lkubb committed:
Rely on timeout fixture for testing infinite loop

lkubb committed:
Merge branch 'master' into script-engine-onchange

lkubb committed:
Add autospec to tests

lkubb committed:
Update versionadded

lkubb committed:
Improve test reliability

* Sometimes, the timeout fails
* Ensure that there are enough values in side_effect for small
  variations

Ch3LL commented:
I noticed more commits. Is this ready for review/merge now? or are you still waiting on https://github.com/saltstack/salt/pull/62910 ?

lkubb commented:
I would consider this ready for review/merging now since I implemented a basic version of the timeout fixture @waynew proposed myself and reverted all the testing-related refactoring of the script engine.

I'm not 100% certain the timeout fixture can be relied on though, in my testing it sometimes did not trigger. This PR contains a workaround for that issue specific to the script engine.

Ch3LL commented:
thanks, just want @waynew to re-review here
